File: public/PublicMedia/static/solutions/elixir/problem3.ex
# Purpose: Find the largest prime factor of the number 600851475143

# Author: Eben Quenneville
# Date: 2023-03-02
import ExUnit.Assertions

defmodule Problem3 do
  # Simpler interface to the main function factors
  def factors_for(n) do
    factors(n, 2, [])
  end

  # Returns the prime factors of the number recursively
  def factors(number, lower_bound, primes) do
    cond do
      lower_bound * lower_bound > number ->
        Enum.reverse([number | primes])

      rem(number, lower_bound) == 0 ->
        factors(div(number, lower_bound), lower_bound, [lower_bound | primes])

      true ->
        factors(number, lower_bound + 1, primes)
    end
  end

  def problem() do
    List.last(factors_for(600_851_475_143))
  end

  # Check solution
  def main() do
    solution = problem()
    assert solution == 6857
    IO.puts("Output: #{solution}")
  end
end

Problem3.main()

File: public/PublicMedia/static/solutions/clojure/problem3.clj
(comment
  "Purpose: Find the largest prime factor of the number 600851475143 
   Author: Eben Quenneville
   Date: 2023-03-04")

(defn factors "Recursively returns the factors of n, bounded by lower" [n lower primes]
  (cond
    (> (* lower lower) n) (reverse (conj primes n))
    (== (rem n lower) 0) (factors (/ n lower) lower (conj primes lower))
    :else (factors n (+ lower 1) primes)))

(defn factors_for "Returns the factors for n as a vector" [n]
  (factors n 2 []))

;; Main driver code
(defn problem3 []
  (apply max (factors_for 600851475143)))

;; Assert the solution
(defn main []
  (assert (== (problem3) 6857))
  (println "Output: " (problem3)))
(main)
